src/detect-base64-decode.c

```diff
@@ -0,0 +1,733 @@
+    "\\s*(\\w+)?";
+static pcre *decode_pcre = NULL;
+static pcre_extra *decode_pcre_study = NULL;
+
+static int DetectBase64DecodeSetup(DetectEngineCtx *, Signature *, char *);
+static void DetectBase64DecodeFree(void *);
+static void DetectBase64DecodeRegisterTests(void);
+
+void DetectBase64DecodeRegister(void)
+{
+    const char *pcre_errptr;
+    int pcre_erroffset;
+
+    sigmatch_table[DETECT_BASE64_DECODE].name = "base64_decode";
+    sigmatch_table[DETECT_BASE64_DECODE].desc =
+        "Decodes base64 encoded data.";
+    sigmatch_table[DETECT_BASE64_DECODE].url =
+        "https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Payload_keywords#base64_decode";
+    sigmatch_table[DETECT_BASE64_DECODE].Setup = DetectBase64DecodeSetup;
+    sigmatch_table[DETECT_BASE64_DECODE].Free = DetectBase64DecodeFree;
+    sigmatch_table[DETECT_BASE64_DECODE].RegisterTests =
+        DetectBase64DecodeRegisterTests;
+
+    sigmatch_table[DETECT_BASE64_DECODE].flags |= SIGMATCH_PAYLOAD;
+    sigmatch_table[DETECT_BASE64_DECODE].flags |= SIGMATCH_OPTIONAL_OPT;
+
+    decode_pcre = pcre_compile(decode_pattern, 0, &pcre_errptr, &pcre_erroffset,
+        NULL);
+    if (decode_pcre == NULL) {
+        SCLogError(SC_ERR_PCRE_COMPILE, "Failed to compile pattern \"%s\" at"
+            " offset %d: %s", decode_pattern, pcre_erroffset, pcre_errptr);
+        exit(EXIT_FAILURE);
+    }
+
+    decode_pcre_study = pcre_study(decode_pcre, 0, &pcre_errptr);
+    if (pcre_errptr != NULL) {
+        SCLogError(SC_ERR_PCRE_STUDY, "Failed to study pattern \"%s\": %s",
+            decode_pattern, pcre_errptr);
+        exit(EXIT_FAILURE);
+    }
+}
+
+int DetectBase64DecodeDoMatch(DetectEngineThreadCtx *det_ctx, Signature *s,
+    const SigMatch *sm, uint8_t *payload, uint32_t payload_len)
+{
+    DetectBase64Decode *data = (DetectBase64Decode *)sm->ctx;
+    int decode_len;
+
+#if 0
+    printf("Input data:\n");
+    PrintRawDataFp(stdout, payload, payload_len);
+#endif
+
+    if (data->relative) {
+        payload += det_ctx->buffer_offset;
+        payload_len -= det_ctx->buffer_offset;
+    }
+
+    if (data->offset) {
+        if (data->offset >= payload_len) {
+            return 0;
+        }
+        payload = payload + data->offset;
+        payload_len -= data->offset;
+    }
+
+    decode_len = MIN(payload_len, data->bytes);
+
+#if 0
+    printf("Decoding:\n");
+    PrintRawDataFp(stdout, payload, decode_len);
+#endif
+
+    det_ctx->base64_decoded_len = DecodeBase64(det_ctx->base64_decoded,
+        payload, decode_len, 0);
+    SCLogDebug("Decoded %d bytes from base64 data.",
+        det_ctx->base64_decoded_len);
+#if 0
+    if (det_ctx->base64_decoded_len) {
+        printf("Decoded data:\n");
+        PrintRawDataFp(stdout, det_ctx->base64_decoded,
+            det_ctx->base64_decoded_len);
+    }
+#endif
+
+    return det_ctx->base64_decoded_len > 0;
+}
+
+static int ParseUint32(const char *str, uint32_t *value)
```

Comment:
ByteExtractStringUint32 is for strings, so perhaps that? Would need to consider how errors are handled.